### ISSUE TITLE
Problem with  training using Gradient Tapes for  deep GP models

### ISSUE THREAD

**Issue by XiaoyuHy** [GPflow/GPflow#1683]
When I performed training using Gradient Tapes for  deep GP models (https://github.com/FelixOpolka/Deep-Gaussian-Process), the code for one iteration of optimisation is as below:

	for x_batch, y_batch in zip(bch_X, bch_Y):
		with tf.GradientTape(watch_accessed_variables=False) as tape:
			tape.watch(dgp.trainable_variables)
			objective = -model.elbo((x_batch, y_batch))
			gradients = tape.gradient(objective, model.trainable_variables)
		optimizer.apply_gradients(zip(gradients, model.trainable_variables))
		elbos.append(-objective.numpy())

 When I do multiple iterations of optimisations, say 100 iterations, the optimiser may fail  and crash at some iteration, with error being a matrix inversion problem. It means the parameters reach a space which would make the matrix non-invertible.  However, optimisers if they can compute the objective at the start should not fail and crash when they try parameters that fail the objective (they should just keep the current objective and sample somewhere else, as what the Scipy optimiser does. 

Just wondered if there is already any solution for this?

Many thanks.

**Comment by st--** [comment]
Hi @XiaoyuHy, nothing about this is specific to GPflow - this is just the standard behaviour of TensorFlow, so you would be better off opening an issue there.

We also recently open-sourced a deep GP implementation at github.com/secondmind-labs/gpflux/ - I'd be interested to hear if you can reproduce the same matrix inversion failure with our code. If that's the case, please open an issue there!
